### PR TITLE
Implement ticklabeloverflow to improve control over display of tick labels

### PR DESCRIPTION
**Pull request by archmoj**
Addressing #5404 by adding new attribute to `cartesian` axes (and few more places than use cartesian axis e.g. `indicator` and `colorbar`, obviously not `gl3d`).

The options currently implemented are `allow`, `hide past div`, `hide past domain`.

The default value when `ticklabelposition` is set to *inside* is 'hide past domain'.
Otherwise on *category* and *multicategory* axes the default is 'allow'.
In other cases the "new" default (to fix the bug) is 'hide past div'.

This PR also paved the way so that other options such as `push to div`, and `push to domain` could be added later on the road to address #3292.

@plotly/plotly_js

### CODE REVIEW

File: src/plots/cartesian/layout_attributes.js [hunk 1]
```diff
@@ -493,6 +493,23 @@ module.exports = {
...
+        values: [
+            'allow',
+            'hide past div',
+            'hide past domain',
+            // 'push to div',
+            // 'push to domain'
+        ],
+        editType: 'calc',
+        description: [
+            'Determines whether or not the tick labels are drawn when overflown the div or domain.',
```
Comment (alexcjohnson) [inline src/plots/cartesian/layout_attributes.js:507]:
```suggestion
            'Determines how we handle tick labels that would overflow either the graph div or the domain of the axis.',
```

File: src/components/colorbar/attributes.js [hunk 2]
```diff
@@ -136,6 +136,7 @@ module.exports = overrideAll({
     tickvals: axesAttrs.tickvals,
     ticktext: axesAttrs.ticktext,
     ticks: extendFlat({}, axesAttrs.ticks, {dflt: ''}),
+    ticklabeloverflow: extendFlat({}, axesAttrs.ticklabeloverflow, {dflt: 'allow'}),
```
Comment (alexcjohnson) [inline src/components/colorbar/attributes.js:139]: You don't want a `dflt` here, do you?

File: src/plots/cartesian/axes.js [hunk 3]
```diff
@@ -3057,63 +3053,81 @@ axes.drawLabels = function(gd, ax, opts) {
...
+                }
+
+                var t = thisLabel.select('text');
+                if(adjust) {
+                    if(hideOverflow) t.style('opacity', 0); // hidden
+                    /*
+                    else if(pushOverflow) {
+                        if(adjust.indexOf('left') !== -1) t.attr('text-anchor', 'start');
+                        if(adjust.indexOf('right') !== -1) t.attr('text-anchor', 'end');
+                        // more TODO: https://github.com/plotly/plotly.js/issues/3292
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:3107]: You could leave another branch open with this WIP for "push to ..." but I'd rather not have it commented out in the main codebase. Anyway the eventual solution is going to need to be a bit more complex than this: we'll want to push these labels only as far as necessary, and then if doing so causes them to intersect the next labels we should still hide them.

**Comment by alexcjohnson** [comment]
The new test images look great. I don't think we want the indicator mock to change though - if indicator labels overflow the div I think we should still show them.

**Comment by archmoj** [comment]
> The new test images look great. I don't think we want the indicator mock to change though - if indicator labels overflow the div I think we should still show them.

Good call. Done in https://github.com/plotly/plotly.js/pull/5584/commits/0deec98f567d552ab8a354cd4cc2b718e203e9c6.

File: src/components/colorbar/attributes.js (hunk 2 above)
Comment (archmoj) [reply src/components/colorbar/attributes.js]: Good eye. Addressed in https://github.com/plotly/plotly.js/pull/5584/commits/f833ce41f9caafc7faab1ebf00861d573caa5687.

File: src/plots/cartesian/axes.js (hunk 3 above)
Comment (archmoj) [reply src/plots/cartesian/axes.js]: Fixed by https://github.com/plotly/plotly.js/pull/5584/commits/c9eaff26eb1f31ebf10f1ba32c714f61e58106a7.

**Review by alexcjohnson — APPROVED** [review]
Good to go! 💃

**Comment by nicolaskruchten** [comment]
⭐ please merge!

**Pull request merged**